pallene/cfg.lua

```diff
@@ -0,0 +1,276 @@
+        assert(new_loop == current_loop,'unmatched loop ending (impossible)')
+        -- at the end of the loop, gets all indexes at the top 
+        -- and points them with what happened before
+        -- to the same with what happened during the loop
+        local top = current_loop[#current_loop]
+
+        -- the condition is the first node of the body
+        -- TODO: remove this hack
+        -- maybe return the first node of the scope, not the only the current
+        -- maybe propose change in the ir
+        local first_node
+        if #current_node == 1 then
+            if type(current_node[1]) == 'number' then
+                first_node =  {current_node[1]+1}
+            else
+                first_node = current_node
+            end
+        elseif #current_node == 0 then
+            first_node = current_node
+        else
+            error('impossible?')
+        end
+        
+        Set.iforeach(top,function(_,v) 
+            Set.iforeach(first_node,function(_,node)
+                make_edge(self,node,v) 
+            end)
+        end) 
+        Set.iforeach(new_nodes,function(_,v) 
+            Set.iforeach(first_node,function(_,node)
+                make_edge(self,node,v) 
+            end)
+        end) 
+        current_loop[#current_loop] = nil
+
+        
+
+        return first_node,new_loop
+
+    elseif tag == "ir.Cmd.If" then
+        -- condition is an ir.Value and 
+        -- will be extracted to an assignment instruction beforehand
+        local then_nodes, then_loop
+        if cmd.then_ then
+            then_nodes, then_loop = build(self,cmd.then_,current_node,current_loop)
+        else 
+            then_nodes, then_loop = {},{}
+        end
+        local else_nodes, else_loop
+        if cmd.else_ then
+            else_nodes, else_loop = build(self,cmd.else_,current_node,current_loop)
+        else 
+            else_nodes, else_loop = {},{}
+        end
+        assert(then_loop == current_loop,'different loop then at meet (impossible)')
+        assert(else_loop == current_loop,'different loop else at meet (impossible)')
+        return Set.union(then_nodes,else_nodes), current_loop
+    elseif string.match(tag, "^ir%.Cmd%.") then
+        return make_node(self,cmd,current_node,current_loop), current_loop
+    else
+        return current_node,current_loop
+    end
+end
+function CFG.new(code)
+    local new = {}
+    -- entry and exit nodes
+    new.entry = {cmd = false, to={},from={}}
+    new.exit = {cmd = false, to={},from={}}
+    local last_node
+    if code then
+        local current_node,_ = build(new,code,{"entry"},{})
+        last_node = current_node
+    else
+        last_node = {'entry'}
+    end
+    Set.iforeach(last_node,function(_,node) 
+        make_edge(new,"exit",node)
+    end)
+    return new
+end
+
+
+
+-- traverses the CFG applying the function f to each node.
+function CFG.traverse_forward(self,f)
```

Comment:
Don't have arguments named `self`. Either the first argument should be caled `cfg` or the function should be a method  (`CFG:traverse_forward`).